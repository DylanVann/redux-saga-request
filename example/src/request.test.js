import { put, call } from 'redux-saga/effects'

import { createRequest, requestSaga } from './request'

const REQUEST_TYPE = createRequest('TEST')
const payload = { foo: 'bar' }
const payloadError = new Error('test')
const func = [() => payload]
const errorFunc = [() => { throw payloadError }]
const meta = { test: 'test' }

describe('saga request helper', () => {
  it('should create action types', () => {
    expect(createRequest(REQUEST_TYPE)).toMatchSnapshot()
  })

  it('should dispatch started and succeeded actions', () => {
    const generator = requestSaga(REQUEST_TYPE, func, meta)
    let next = generator.next()

    expect(next.value).toEqual(
      put(REQUEST_TYPE.start(meta)),
      'it should dispatch a started action with the meta')

    next = generator.next(REQUEST_TYPE.start(meta))

    expect(next.value).toEqual(
      call(...func),
      'it should call the provided function')

    next = generator.next(payload)

    expect(next.value).toEqual(
      put(REQUEST_TYPE.success(payload, meta)),
      'should dispatch a succeeded action with the payload and meta')

    next = generator.next()

    expect(next.value).toEqual(
      { '@@redux-saga/IO': true, CANCELLED: {} },
      'it should check if the saga was cancelled',
    )

    next = generator.next()

    expect(next.value).toBeUndefined()
    expect(next.done).toEqual(true, 'the generator should have finished')
  })

  it('should dispatch started and errored actions', () => {
    const generator = requestSaga(REQUEST_TYPE, errorFunc, meta)
    let next = generator.next()

    expect(next.value).toEqual(
      put(REQUEST_TYPE.start(meta)),
      'it should dispatch a started action with the meta')

    next = generator.next(REQUEST_TYPE.start(meta))

    expect(next.value).toEqual(
      call(...errorFunc),
      'it should call the provided function')

    next = generator.throw(payloadError)

    expect(next.value).toEqual(
      put(REQUEST_TYPE.error(payloadError, meta)),
      'should dispatch a errored action with the error and meta')

    next = generator.next()

    expect(next.value).toEqual(
      { '@@redux-saga/IO': true, CANCELLED: {} },
      'it should check if the saga was cancelled',
    )

    next = generator.next()

    expect(next.value).toBeUndefined()
    expect(next.done).toEqual(true, 'the generator should have finished')
  })
})
