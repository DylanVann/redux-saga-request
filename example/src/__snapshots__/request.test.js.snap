// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saga request helper should create action types 1`] = `
Object {
  "BASE": Object {
    "BASE": "TEST",
    "CANCELLED": "TEST_CANCELLED",
    "ERRORED": "TEST_ERRORED",
    "STARTED": "TEST_STARTED",
    "SUCCEEDED": "TEST_SUCCEEDED",
    "cancel": [Function],
    "error": [Function],
    "start": [Function],
    "success": [Function],
  },
  "CANCELLED": "[object Object]_CANCELLED",
  "ERRORED": "[object Object]_ERRORED",
  "STARTED": "[object Object]_STARTED",
  "SUCCEEDED": "[object Object]_SUCCEEDED",
  "cancel": [Function],
  "error": [Function],
  "start": [Function],
  "success": [Function],
}
`;
